### PR TITLE
Add php7.4 support

### PR DESCRIPTION
**Pull request by rullzer**
Add php7.4 support

Needs:
- [x] https://github.com/nextcloud/server/pull/18068
- [x] https://github.com/nextcloud/server/pull/18073

### CODE REVIEW

**Comment by rullzer** [comment]
Going better but still some failing stuff

https://drone.nextcloud.com/nextcloud/server/23525/2/4
https://drone.nextcloud.com/nextcloud/server/23525/3/4

**Comment by rullzer** [comment]
Some help in the final fixes is appreciated. As I'd like to merge this sooner than later/.

**Comment by rullzer** [comment]
Yay. Ok lets do this and then I crate PRs for the remaining test:

File: apps/dav/tests/unit/SystemTag/SystemTagNodeTest.php [hunk 1]
```diff
@@ -286,10 +286,10 @@ public function testDeleteTagPermissionException(ISystemTag $tag, $expectedExcep
 		$this->getTagNode(false, $tag)->delete();
 	}
 
-	/**
-	 * @expectedException \Sabre\DAV\Exception\NotFound
-	 */
+	
 	public function testDeleteTagNotFound() {
+	    $this->expectException(\Sabre\DAV\Exception\NotFound::class);
```
Comment (ChristophWurst) [inline apps/dav/tests/unit/SystemTag/SystemTagNodeTest.php]: that indentation :roll_eyes:

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
The indentation of $this->expectException is consistently inconsistent

File: apps/dav/tests/unit/SystemTag/SystemTagNodeTest.php (hunk 1 above)
Comment (rullzer) [reply apps/dav/tests/unit/SystemTag/SystemTagNodeTest.php]: ah php-cs-fixer setting... fixed.

**Review by ChristophWurst — APPROVED** [review]

**Review by blizzz — APPROVED** [review]

**Review by juliusknorr — APPROVED** [review]

**Comment by Thaodan** [comment]
Will this be added to the next point release of 17.x?

**Comment by nickvergessen** [comment]
Unlikely. We try to keep the same php versions for a major once it is released

**Comment by Thaodan** [comment]
Doesn't sound good for distros that feature php 7.4
I removed the version check and currently everything is fine.

Would these changes break older versions?

**Comment by rullzer** [comment]
They could since we don't test them. 18 is set to be released mid January which will support it.

**Comment by nursoda** [comment]
So, at least on Arch I don not have PHP 7.3 and 7.4 to cheese from, it's just 'current'. So I may either not patch PHP (and related packages) for >6 weeks (not a good idea IF security relevant patches to PHP appear, apart from dependency quirks), or run NC 17 on an not only untested but also unsupported PHP version hoping that changes merely create warnings.

I would be interested to hear which other major distros have such issue. If non, then I agree that Arch was not be a good choice for a production system. But on the other hand, PHP 7.4 did not just pop up yesterday and compatibility checks or could have been done meanwhile, or even adaption of current NC stable, not only NC next.

**Comment by go2sh** [comment]
@nursoda Most other Distro allow multiple php Versions to be installed side-by-side and the packages are called php-7.0, php-7.1 and so on. For normal release distros, you can install the distro version and some other version (backport, custom repo). Arch does this for python, but there are many other packages, which have the same problem, e.g. the kernel.

**Comment by Thaodan** [comment]
> (backport, custom repo). Arch does this for python, but there are many other 
No that different because python2 and python3 are incompatible. Its like php 5 
and 7.

Usally its not that good to have multiple minor versions of software if its 
aviodable.
I think distros should backport fixes if the software needs fixes.

**Comment by alerque** [comment]
7.3 → 7.4 is not a major version change that distros should be expected to check every webapp against. The latest 7.x should be workable and it is incumbent on projects like this one to support it them. Distros shouldn't have to backport a bunch of patches to run on the latest interpreter, if anything only the opposite scenario should be necessary.

**Comment by rullzer** [comment]
@alerque as said we will support it with 18 that is scheduled for mid january.

@nursoda debian/ubuntu, red hat/centos, suse etc all do not feature php 7.4 yet and usually stick to a version on a new release. Or often they allow you to install other versions. Gentoo is similar and also doesn't have 7.4 in stable yet.

So, as said you can remove the version check at your own risk. Or you wait till end of the week to try out the beta of 18.

**Comment by blizzz** [comment]
@alerque PHP 7.4 (as any other releases of this level) bring their own backward incompatibilities: https://www.php.net/manual/en/migration74.incompatible.php

**Comment by eli-schwartz** [comment]
> @nursoda Most other Distro allow multiple php Versions to be installed side-by-side and the packages are called php-7.0, php-7.1 and so on.

I took a look at debian as an example: https://packages.debian.org/search?suite=all&searchon=names&keywords=php

I see exactly one version of php available in each of Debian's release channels. Stretch has `php7.0` exclusively, Buster/Bullseye/Sid have `php7.3` exclusively, Experimental has `php7.4` exclusively.

Arch Linux has `php` exclusively, currently at version `7.4`.

Please don't try to pretend that a distro supporting one version of php at a time is somehow bizarre.

**Comment by kesselb** [comment]
What's the intention of your post? Nobody said that "a distro supporting one version of php at a time is somehow bizarre". That's your interpretation. A user shared his way to deal with this kind of problem. Nothing wrong with that. I think there is a discussion at help.nextcloud.com regarding release strategies and early support for PHP 7.4. This pull request adds support for PHP 7.4 and is the wrong place to discuss such things.

**Comment by go2sh** [comment]
> Please don't try to pretend that a distro supporting one version of php at a time is somehow bizarre.
I never said that and you have quoted only the first part. I said you need some other source like backports or custom repos to install any other version then the current distro version. But every php package follow a certain naming schema and directory layout, which makes it easily possible to install multiple version side by side.

And its the same with arch. Official repo 7.4, AUR what ever php version there is. One user in the bug tracker said the right thing:
> Perhaps Nextcloud should be pushed back into to AUR where having multiple versions of a package is more palatable? Upgrading right away to NC18 might not always be desirably anyway as plugins tend to lag behind, so multiple versions of nextcloud and multiple versions of php in AUR would probably offer a better user experience.

**Comment by Thaodan** [comment]
I don't sse how the AUR multiple version better 
than the community repo where it currently is in.
Usally there's a check for dependency issues 
when there's a new version of a dependency.
Maybe packagers should check for the exact 
version just as nextcloud does.

**Comment by ChristophWurst** [comment]
Let's discuss this in a ticket or on help.nextcloud.com.

**Pull request merged**